File: python/problem_49.py
from typing import List


def max_subarray_sum(arr: List[int]) -> int:
    best = cur = 0
    for v in arr:
        cur = max(0, cur + v)
        best = max(best, cur)
    return best


def main():
    print("Problem 49:")
    assert max_subarray_sum([34, -50, 42, 14, -5, 86]) == 137
    assert max_subarray_sum([-5, -1, -8, -9])          == 0
    assert max_subarray_sum([1, 2, 3])                 == 6
    print("All Tests Passed")

File: go/problem_49.go
package main

func maxSubarraySum(arr []int) int {
	best, cur := 0, 0
	for _, v := range arr {
		cur += v
		if cur > best {
			best = cur
		}
		if cur < 0 {
			cur = 0
		}
	}
	return best
}

func problem49() {
	println("Problem 49:")
	assertEqual(maxSubarraySum([]int{34, -50, 42, 14, -5, 86}), 137)
	assertEqual(maxSubarraySum([]int{-5, -1, -8, -9}), 0)
	assertEqual(maxSubarraySum([]int{1, 2, 3}), 6)
	println("All Tests Passed")
}
